src/server/src/main/java/io/cassandrareaper/service/RepairRunner.java

```diff
@@ -272,10 +282,108 @@ private void endRepairRun() {
+   */
+  @VisibleForTesting
+  public void maybeAdaptRepairSchedule() {
+    Optional<RepairRun> repairRun = context.storage.getRepairRun(repairRunId);
+    if (repairRun.isPresent() && Boolean.TRUE.equals(repairRun.get().getAdaptiveSchedule())) {
+      Collection<RepairSegment> segments
+          = context.storage.getSegmentsWithState(repairRunId, RepairSegment.State.DONE);
+      int maxSegmentDurationInMins = segments.stream()
+          .mapToInt(repairSegment
+              -> (int) (repairSegment.getEndTime().getMillis() - repairSegment.getStartTime().getMillis()) / 60_000)
+          .max()
+          .orElseThrow(NoSuchElementException::new);
+      int extendedSegments = (int) segments.stream().filter(segment -> segment.getFailCount() > 0).count();
+      double percentExtendedSegments
+          = ((float) extendedSegments / (float) repairRun.get().getSegmentCount()) * 100.0;
+      LOG.info("extendedSegments = {}, total segments = {}, percent extended = {}",
+          extendedSegments,
+          repairRun.get().getSegmentCount(),
+          percentExtendedSegments);
+      tuneAdaptiveRepair(percentExtendedSegments, maxSegmentDurationInMins);
+    }
+  }
+
+  @VisibleForTesting
+  void tuneAdaptiveRepair(double percentExtendedSegments, int maxSegmentDuration) {
+    LOG.info("Percent extended segments: {}", percentExtendedSegments);
+    if (percentExtendedSegments > PERCENT_EXTENDED_THRESHOLD) {
+      // Too many segments were extended, meaning that the number of segments is too low
+      addSegmentsPerNodeToScheduleForUnit();
+    } else if ((int) percentExtendedSegments <= PERCENT_EXTENDED_THRESHOLD && percentExtendedSegments >= 1) {
+      // The number of extended segments is moderate and timeout should be extended
+      raiseTimeoutOfUnit();
+    } else if (percentExtendedSegments == 0 && maxSegmentDuration < SEGMENT_DURATION_FOR_REDUCTION_THRESHOLD) {
+      // Segments are being executed very fast and could be lowered
+      reduceSegmentsPerNodeToScheduleForUnit();
+    }
+  }
+
+  @VisibleForTesting
+  void reduceSegmentsPerNodeToScheduleForUnit() {
+    LOG.debug("Reducing segments per node for adaptive schedule on repair unit {}", repairUnit.getId());
+    RepairSchedule scheduleToTune = getScheduleForRun();
+
+    // Reduce segments by decrements of 10%
+    int newSegmentCountPerNode
+        = (int) Math.max(MIN_SEGMENTS_PER_NODE_REDUCTION, scheduleToTune.getSegmentCountPerNode() / 1.1);
+
+    // update schedule with new segment per node value
+    RepairSchedule newSchedule
+        = scheduleToTune.with().segmentCountPerNode(newSegmentCountPerNode).build(scheduleToTune.getId());
+    context.storage.updateRepairSchedule(newSchedule);
+
+  }
+
+  @VisibleForTesting
+  void raiseTimeoutOfUnit() {
+    // Build updated repair unit
+    RepairUnit updatedUnit = repairUnit.with().timeout(repairUnit.getTimeout() * 2).build(repairUnit.getId());
```

Comment:
doubling is a big jump. any evidence/metrics on why we are choosing this? my guess would have been a 50% (but i really don't know)

Reply:
I'd say it's not a big risk as what can happen at worse is that the segment will take more time to be canceled if it gets stuck. I'd rather have that than multiple failed attempts as the odds of having stuck segments lowered a lot with the latest versions of Cassandra.
wdyt?

Reply:
wfm